pkg/scheduler/framework/plugins/dynamicresources has loopvar sharing bug

### What happened?

Apply this patch:

```
diff --git a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
index 5d09769aa39..7043725c456 100644
--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
@@ -589,11 +589,13 @@ func (tc *testContext) listAll(t *testing.T) (objects []metav1.Object) {
 	claims, err := tc.client.ResourceV1alpha2().ResourceClaims("").List(tc.ctx, metav1.ListOptions{})
 	require.NoError(t, err, "list claims")
 	for _, claim := range claims.Items {
+		claim := claim
 		objects = append(objects, &claim)
 	}
 	schedulings, err := tc.client.ResourceV1alpha2().PodSchedulingContexts("").List(tc.ctx, metav1.ListOptions{})
 	require.NoError(t, err, "list pod scheduling")
 	for _, scheduling := range schedulings.Items {
+		scheduling := scheduling
 		objects = append(objects, &scheduling)
 	}
 
```

Then run:

```
make test WHAT=./pkg/scheduler/framework/plugins/dynamicresources
```

This fails with:

```
--- FAIL: TestPlugin (0.00s)
    --- FAIL: TestPlugin/claim-template (0.11s)
        dynamicresources.go:323: I0521 00:50:10.956189] pod resource claims pod="default/my-pod" resourceclaims=["default/my-pod-my-resource"]
        dynamicresources.go:579: I0521 00:50:10.958622] all potential nodes already set pod="default/my-pod" potentialnodes=[{"metadata":{"name":"worker","creationTimestamp":null,"labels":{"nodename":"worker"}},"spec":{},"status":{"capacity":{"pods":"32"},"allocatable":{"pods":"32"},"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}]
        dynamicresources.go:635: I0521 00:50:10.959354] reserve pod="default/my-pod" node="worker" resourceclaim="default/my-pod-my-resource"
        --- FAIL: TestPlugin/claim-template/reserve (0.00s)
            dynamicresources_test.go:501: 
                	Error Trace:	/usr/local/google/home/rsc/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:566
                	            				/usr/local/google/home/rsc/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:501
                	Error:      	Not equal: 
                	            	expected: []v1.Object{(*v1alpha2.ResourceClaim)(0xc0004faa80), (*v1alpha2.ResourceClaim)(0xc0004faf00)}
                	            	actual  : []v1.Object{(*v1alpha2.ResourceClaim)(0xc0004fbe00), (*v1alpha2.ResourceClaim)(0xc000936000)}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -50,3 +50,10 @@
                	            	    }),
                	            	-   ReservedFor: ([]v1alpha2.ResourceClaimConsumerReference) <nil>,
                	            	+   ReservedFor: ([]v1alpha2.ResourceClaimConsumerReference) (len=1) {
                	            	+    (v1alpha2.ResourceClaimConsumerReference) {
                	            	+     APIGroup: (string) "",
                	            	+     Resource: (string) (len=4) "pods",
                	            	+     Name: (string) (len=6) "my-pod",
                	            	+     UID: (types.UID) (len=4) "1234"
                	            	+    }
                	            	+   },
                	            	    DeallocationRequested: (bool) false
                	Test:       	TestPlugin/claim-template/reserve
        dynamicresources.go:773: I0521 00:50:10.961401] no PodSchedulingContext object to delete
    --- FAIL: TestPlugin/claim-reference (0.11s)
        dynamicresources.go:323: I0521 00:50:10.957592] pod resource claims pod="default/my-pod" resourceclaims=["default/my-pod-my-resource"]
        dynamicresources.go:579: I0521 00:50:10.959183] all potential nodes already set pod="default/my-pod" potentialnodes=[{"metadata":{"name":"worker","creationTimestamp":null,"labels":{"nodename":"worker"}},"spec":{},"status":{"capacity":{"pods":"32"},"allocatable":{"pods":"32"},"daemonEndpoints":{"kubeletEndpoint":{"Port":0}},"nodeInfo":{"machineID":"","systemUUID":"","bootID":"","kernelVersion":"","osImage":"","containerRuntimeVersion":"","kubeletVersion":"","kubeProxyVersion":"","operatingSystem":"","architecture":""}}}]
        dynamicresources.go:635: I0521 00:50:10.959823] reserve pod="default/my-pod" node="worker" resourceclaim="default/my-pod-my-resource"
        --- FAIL: TestPlugin/claim-reference/reserve (0.00s)
            dynamicresources_test.go:501: 
                	Error Trace:	/usr/local/google/home/rsc/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:566
                	            				/usr/local/google/home/rsc/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:501
                	Error:      	Not equal: 
                	            	expected: []v1.Object{(*v1alpha2.ResourceClaim)(0xc0006bc900), (*v1alpha2.ResourceClaim)(0xc0006bca80)}
                	            	actual  : []v1.Object{(*v1alpha2.ResourceClaim)(0xc0006bd980), (*v1alpha2.ResourceClaim)(0xc0006bdb00)}
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -50,3 +50,10 @@
                	            	    }),
                	            	-   ReservedFor: ([]v1alpha2.ResourceClaimConsumerReference) <nil>,
                	            	+   ReservedFor: ([]v1alpha2.ResourceClaimConsumerReference) (len=1) {
                	            	+    (v1alpha2.ResourceClaimConsumerReference) {
                	            	+     APIGroup: (string) "",
                	            	+     Resource: (string) (len=4) "pods",
                	            	+     Name: (string) (len=6) "my-pod",
                	            	+     UID: (types.UID) (len=4) "1234"
                	            	+    }
                	            	+   },
                	            	    DeallocationRequested: (bool) false
                	Test:       	TestPlugin/claim-reference/reserve
        dynamicresources.go:773: I0521 00:50:10.961672] no PodSchedulingContext object to delete
FAIL
```



### What did you expect to happen?

Test should pass.

### How can we reproduce it (as minimally and precisely as possible)?

See above.

### Anything else we need to know?

If there are N objects in claims.Items, then this loop

	for _, claim := range claims.Items {
		objects = append(objects, &claim)
	}

is appending the same pointer N times, pointing at the final item in the list. Adding `claim := claim` fixes that mistake, but then it makes the test fail. Something else about the test is buggy, but the bug was being masked by this bug.

In the repro case I added `scheduling := scheduling` to the loop a few lines down, just for good measure, but that one is not causing a problem. The `claim := claim` line by itself causes the test failure.



### Kubernetes version

HEAD

### Cloud provider

N/A

### OS version

N/A

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

--- a/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
+++ b/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go
@@ -104,7 +104,7 @@ var (
 				Obj()
 	inUseClaim = st.FromResourceClaim(pendingImmediateClaim).
 			Allocation(&resourcev1alpha2.AllocationResult{}).
-			ReservedFor(resourcev1alpha2.ResourceClaimConsumerReference{UID: types.UID(podUID)}).
+			ReservedFor(resourcev1alpha2.ResourceClaimConsumerReference{Resource: "pods", Name: podName, UID: types.UID(podUID)}).
 			Obj()
 	allocatedClaim = st.FromResourceClaim(pendingDelayedClaim).
 			OwnerReference(podName, podUID, podKind).
@@ -214,10 +214,36 @@ func TestPlugin(t *testing.T) {
 		"claim-reference": {
 			pod:    podWithClaimName,
 			claims: []*resourcev1alpha2.ResourceClaim{allocatedClaim, otherClaim},
+			want: want{
+				reserve: result{
+					changes: change{
+						claim: func(claim *resourcev1alpha2.ResourceClaim) *resourcev1alpha2.ResourceClaim {
+							if claim.Name == claimName {
+								claim = claim.DeepCopy()
+								claim.Status.ReservedFor = inUseClaim.Status.ReservedFor
+							}
+							return claim
+						},
+					},
+				},
+			},
 		},
 		"claim-template": {
 			pod:    podWithClaimTemplate,
 			claims: []*resourcev1alpha2.ResourceClaim{allocatedClaim, otherClaim},
+			want: want{
+				reserve: result{
+					changes: change{
+						claim: func(claim *resourcev1alpha2.ResourceClaim) *resourcev1alpha2.ResourceClaim {
+							if claim.Name == claimName {
+								claim = claim.DeepCopy()
+								claim.Status.ReservedFor = inUseClaim.Status.ReservedFor
+							}
+							return claim
+						},
+					},
+				},
+			},
 		},
 		"missing-claim": {
 			pod: podWithClaimTemplate,
@@ -589,11 +615,13 @@ func (tc *testContext) listAll(t *testing.T) (objects []metav1.Object) {
 	claims, err := tc.client.ResourceV1alpha2().ResourceClaims("").List(tc.ctx, metav1.ListOptions{})
 	require.NoError(t, err, "list claims")
 	for _, claim := range claims.Items {
+		claim := claim
 		objects = append(objects, &claim)
 	}
 	schedulings, err := tc.client.ResourceV1alpha2().PodSchedulingContexts("").List(tc.ctx, metav1.ListOptions{})
 	require.NoError(t, err, "list pod scheduling")
 	for _, scheduling := range schedulings.Items {
+		scheduling := scheduling
 		objects = append(objects, &scheduling)
 	}
 

